File: src/arduinobot_cpp_examples/src/simple_service_client.cpp
#include <rclcpp/rclcpp.hpp>
#include <arduinobot_msgs/srv/add_two_ints.hpp>
#include <memory>
#include <chrono>

using namespace std::chrono_literals;
using std::placeholders::_1;

class SimpleServiceClient: public rclcpp::Node{
    public:
    SimpleServiceClient(int a, int b) : Node("simple_service_client"){
        client_ = create_client<arduinobot_msgs::srv::AddTwoInts>("add_two_ints");
        
        auto request = std::make_shared<arduinobot_msgs::srv::AddTwoInts::Request>();
        request->a = a;
        request->b = b;

        while (!client_->wait_for_service(1s))
        {
            RCLCPP_ERROR(get_logger(), "Waiting for service");
            if(!rclcpp::ok()){
                RCLCPP_ERROR(get_logger(), "Interrupted while waiting for service to start");
                return;
            }
        }

        auto result = client_->async_send_request(request, std::bind(&SimpleServiceClient::responseCallback, this, _1));
    }

    private:
    rclcpp::Client<arduinobot_msgs::srv::AddTwoInts>::SharedPtr client_;

    void responseCallback(rclcpp::Client<arduinobot_msgs::srv::AddTwoInts>::SharedFuture future){
        if (future.valid()){
            RCLCPP_INFO_STREAM(get_logger(), "Response: " << future.get()->sum);
            return;
        }
        RCLCPP_ERROR(get_logger(), "Invalid Response from the service");
    }

};

int main(int argc, char* argv[]){
    if(argc != 3)
    {
        RCLCPP_ERROR(rclcpp::get_logger("rclcpp"), "Wrong number of arguments");
        return -1;
    }
    rclcpp::init(argc, argv);
    auto simpleServiceClient = std::make_shared<SimpleServiceClient>(atoi(argv[1]),atoi(argv[2]));
    rclcpp::spin(simpleServiceClient);
    rclcpp::shutdown();
    return 0;
}

File: src/arduinobot_py_examples/arduinobot_py_examples/simple_service_client.py
import rclpy
from rclpy.node import Node
from arduinobot_msgs.srv import AddTwoInts
import sys

class SimpleServiceClient(Node):
    def __init__(self, a, b):
        super().__init__('simple_service_client')
        
        self.client_ = self.create_client(AddTwoInts, 'add_two_ints')

        while not self.client_.wait_for_service(timeout_sec=1.0):
            self.get_logger("Waiting for add_two_ints service to be ready...")

        self.req = AddTwoInts.Request()
        self.req.a = a
        self.req.b = b

        self.future_ = self.client_.call_async(self.req)

        self.future_.add_done_callback(self.responseCallback)

    def responseCallback(self, future):
        print(future.result())
        sum = future.result().sum
        self.get_logger().info(f'Response: {sum}')



def main():
    if len(sys.argv) != 3:
        print("Wrong number of arguments")
        return -1
        
    rclpy.init()
    simpleServiceClient = SimpleServiceClient(int(sys.argv[1]),int(sys.argv[2]))
    rclpy.spin(simpleServiceClient)
    simpleServiceClient.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()